### ISSUE TITLE
Recursive Relations hang indefinitely

### ISSUE THREAD

**Issue by TylerR909** [joist-orm/joist-orm#1138]
I'm running into a scenario where a recursive relation (at the hint level, not necessarily that the entities are actually recursive) hangs indefinitely when `em.refresh({ forceReload: true })` is called by Tests.

So let's say an `Author` has many `stories` but some of those `stories` are either just 1 story or a collection of other stories. So the data might look like:

```
story1: type=one
story2: type=one
story3: type=anthology, stories:[story1]
story4: type=anthology, stories[story2, story3]
```

Let's say I wanted to write a Collection to "flatmap" all my stories together. That might look like:

```js
class Stories {
  readonly allStories: Collection<Stories, Stories> = hasManyDerived(
    { stories: "allStories" }, // <-- recursive relationship even though none of the data is recursive
    {
      get: (story) => {
        if (story.type === "one") return [story];
        return story.stories.get.flatMap((story) => story.allStories.get);
      },
    }
);
```

So far this actually works exactly as desired. Entity tests pass. Flush flushes. It's honestly working pretty well. But there is one scenario in which it breaks:

1. Given: a recursive load hint 
2. EntityManager must have 2+ `stories` in scope 
3. `.load({ forceRefresh: true })` must be called on the relation/collection
4. They must be called in parallel

GraphQL Resolver tests using the `run()/makeRun()` helpers set up a situation in which `em.refresh()` gets called*. `refresh` invokes `await Promise.all(custom.map((r) => r.load({ forceReload: true })));` which stalls and never recovers. If I change it to a `for(const...of)` (serialize it instead of parallel) it works. If I call it without `forceReload` it works. And if there's only 1 `story` in scope it works.

What I suspect is happening is the Promise.all effectively invokes:

```
GIVEN (simplified):
story1: type=one
story2: type=anthology, stories:[story1]

THEN refresh will:
...load story1 allStories
...load story2 allStories (...load story1 allStories)
```

And that sets up some sort of deadlock or mis-matched await that never recovers.

\* - I can't reproduce it directly with `em.refresh({ deepLoad: true })` but I can reproduce it manually with:

```js
const story1 = newStory(em, { type: One });
const story2 = newStory(em, { type: Anthology, stories: [story1] });

await Promise.all([
  story1.allStories.load({ forceReload: true }),
  story2.allStories.load({ forceReload: true })
]); // jest timeout
```

As you can see this has all the constituent elements: The recursive hint, 2+ entities, called in parallel, with force. Maybe `refresh` is doing more filtering that I realize and GraphQL dirties entities enough to trigger it, but doing it by hand also works.

Do you think this is fixable?

**Comment by stephenh** [comment]
Oh, I am more thoroughly reading the use case now...

Right, I'm actually surprised this scenario:

```
GIVEN (simplified):
story1: type=one
story2: type=anthology, stories:[story1]
```

Works as well as it does, i.e. it works until `em.refresh`.

This is a recursive relationship, i.e. parent `Story`s can have child `Story`s which can have grandchild `Story`s etc.

With `allStories` as written, Joist will issue 1 SQL per level of stories, i.e. if there is a story with 10 levels of grandchildren, we're going to do 10 SQL queries continually going down the tree, until we don't find any children.

This is fine-ish, i.e. probably something we can ship (assuming the `em.refresh` bug is fixed), but ideally we could use a recursive Common Table Expression:

```
WITH RECURSIVE PartsHierarchy AS (
    -- Anchor member: Select the root parts (those without a parent)
    SELECT 
        id,
        name,
        parent_id,
        1 AS level
    FROM 
        parts
    WHERE 
        parent_id IS NULL

    UNION ALL

    -- Recursive member: Select the child parts and join with the recursive CTE
    SELECT 
        p.id,
        p.name,
        p.parent_id,
        ph.level + 1 AS level
    FROM 
        parts p
    INNER JOIN 
        PartsHierarchy ph ON p.parent_id = ph.id
)
-- Select from the CTE to get the hierarchy
SELECT 
    id,
    name,
    parent_id,
    level
FROM 
    PartsHierarchy
ORDER BY 
    level, id;
```

This would bring back all 10 levels of children/grand-children/great-etc in a single call.

It seems pretty doable to get Joist to issue the ^ query, the biggest wrinkle is how it integrates with the "WIP changes" in the `EntityManager`. I.e. a big feature of the `EntityManager` is how both sides of a relation are always in sync; we implicitly rely on it a lot. But with this "go find 10 levels of children", ideally we'd want to make sure a mutation that, say, the 5th level down (maybe a new story added, or a story removed), is immediately reflected in the `allStories` result.

**Comment by stephenh** [comment]
@TylerR909 we have recursive relations now, so can you use `story.storiesRecursive.get` for this use case, instead of the hand-written `allStories`?

**Comment by stephenh** [comment]
Closing as fixed in the recursive relation work.
